MennaMohie/assignment2/Solution.cpp

```diff
@@ -12,118 +14,111 @@ class Node
 
 public:
-    Node():left(nullptr),right(nullptr) {}
     Node(T val):left(nullptr),right(nullptr),value(val) {}
 
-    T getValue(){return value;}
+    T getValue()
+    {
+        return value;
+    }
 };
 
 template<class T>
 class BinaryTree
 {
     Node<T> *root;
-
-public:
-    BinaryTree(T val):root(new Node<T>(val)) {}
-
-    Node<T>* getRoot()
-    {
-        return root;
-    }
-
-    bool contains(Node<T> *currentRoot, T val)
-    {
-        if(currentRoot == nullptr)
-            return false;
-        if(currentRoot->value == val)
-            return true;
-        return (contains(currentRoot->left, val) || contains(currentRoot->right, val));
-    }
-
-    bool resetRoot(T val)
+    int Size;
+    add(Node<T>* currentRoot, T parentVal, T val, direction dir)
     {
-        if(root==nullptr)
-        {
-            root=new Node<T>(val);
-            return true;
-        }
-        else if(!contains(root, val))
-        {
-            root->value=val;
-            return true;
-        }
-        else
-            return false;
-    }
-
-    //adds a new node (val) as a left child to an existing node (parentVal)
-    bool addAsLeftChild(Node<T> *currentRoot, T parentVal, T val)
-    {
-        if(!contains(root, parentVal) || contains(root, val) || currentRoot==nullptr)
-            return false;
         if(currentRoot->value==parentVal)
         {
-            if(currentRoot->left==nullptr)
+            if(dir==0)
             {
-                currentRoot->left=new Node<T>(val);
-                return true;
+                if(currentRoot->left==nullptr)
+                {
+                    currentRoot->left=new Node<T>(val);
+                    Size++;
+                    return true;
+                }
+                else
+                    return false;
             }
-            else
+            if(dir==1)
             {
-                return false;
+                if(currentRoot->right==nullptr)
+                {
+                    currentRoot->right=new Node<T>(val);
+                    return true;
+                }
+                else
+                    return false;
             }
         }
-        if(addAsLeftChild(currentRoot->left, parentVal, val))
-            return true;
-        return addAsLeftChild(currentRoot->right, parentVal, val);
+        if(dir==0)
+        {
+            if(addChildToParent(currentRoot->left, parentVal, val, addLeft))
+                return true;
+            return addChildToParent(currentRoot->right, parentVal, val, addLeft);
+        }
+        if(dir==1)
+        {
+            if(addChildToParent(currentRoot->left, parentVal, val, addRight))
+                return true;
+            return addChildToParent(currentRoot->right, parentVal, val, addRight);
+        }
+    }
+public:
+    BinaryTree(T val):root(new Node<T>(val)),Size(1) {}
+
+    Node<T>* getRoot()
+    {
+        return root;
+    }
 
+    Node<T>* contains(Node<T> *currentRoot, T val)
```

Comment:
NIT: function name suggest just a bool return indicating if the node is in the tree.
For a function actually returning the node something like findNode, getNode, etc. will be better.